### PR TITLE
fix(layout): scale the page-number bands by the page's real height

### LINKED ISSUE #283: Page-number stripping deletes body content on A4 pages (PAGE_NUMBER_TOP_Y is US-Letter-calibrated)
## Summary

`page_number_value` treats any short all-digit text item near the top or bottom of a page as a page-number candidate, using the **absolute** constant `PAGE_NUMBER_TOP_Y = 720.0`. That constant is never scaled by the actual page height.

On US Letter (792 pt) the top band is 72 pt — one inch, as intended. On **A4 (841.89 pt) it is the top 121.89 pt, roughly 14.5% of the page**, or about eight lines of body text — even though the comment beside the constant says "top ~5%".

The result is that body content gets silently deleted from A4 documents. The output stays well-formed, so the loss is easy to miss: a table keeps a valid pipe structure and simply loses a column.

## Minimal reproduction

Self-contained, no external files. Same two rows, drawn twice on an A4 page — once at y=740/725, once at y=700/685:

```python
"""Minimal repro: same two rows, once at y=740 and once at y=700, on A4."""
import pdf_inspector as pi

def make_pdf(path, rows):
    lines = []
    for y, idx, txt in rows:
        lines.append(f"BT /F1 9 Tf 208 {y} Td ({idx}) Tj ET")
        lines.append(f"BT /F1 9 Tf 226 {y} Td ({txt}) Tj ET")
    content = "\n".join(lines).encode()
    objs = [
        b"<< /Type /Catalog /Pages 2 0 R >>",
        b"<< /Type /Pages /Kids [3 0 R] /Count 1 >>",
        b"<< /Type /Page /Parent 2 0 R /MediaBox [0 0 595.28 841.89] "
        b"/Resources << /Font << /F1 5 0 R >> >> /Contents 4 0 R >>",
        b"<< /Length %d >>\nstream\n" % len(content) + content + b"\nendstream",
        b"<< /Type /Font /Subtype /Type1 /BaseFont /Helvetica >>",
    ]
    out = bytearray(b"%PDF-1.4\n"); offs = []
    for i, o in enumerate(objs, 1):
        offs.append(len(out)); out += b"%d 0 obj\n" % i + o + b"\nendobj\n"
    x = len(out)
    out += b"xref\n0 %d\n0000000000 65535 f \n" % (len(objs) + 1)
    for o in offs: out += b"%010d 00000 n \n" % o
    out += b"trailer\n<< /Size %d /Root 1 0 R >>\nstartxref\n%d\n%%%%EOF\n" % (len(objs) + 1, x)
    open(path, "wb").write(bytes(out))

for label, base in (("TOP  (y=740/725)", 740), ("BODY (y=700/685)", 700)):
    make_pdf("r.pdf", [(base, "1", "22 December 2026"), (base - 15, "2", "22 January 2027")])
    print(f"{label}: {pi.process_pdf('r.pdf').markdown.strip()!r}")
```

Output:

```
TOP  (y=740/725): '22 December 2026 22 January 2027'
BODY (y=700/685): '1 22 December 2026 2 22 January 2027'
```

The indices `1` and `2` are deleted in the first case and kept in the second. Only the y offset differs. A 1 pt bisection on a real document puts the boundary exactly at 720: at `y = 720.78` the index is lost, at `y = 719.78` it survives. Changing the MediaBox height does not move the boundary, confirming the threshold is absolute.

## Root cause

`src/extractor/layout.rs:973-995`:

```rust
const PAGE_NUMBER_BOTTOM_Y: f32 = 100.0;
const PAGE_NUMBER_TOP_Y: f32 = 720.0;

fn page_number_value(item: &TextItem) -> Option<u32> {
    // ...
    let text = item.text.trim();
    if text.is_empty() || text.len() > 4 || !text.chars().all(|c| c.is_ascii_digit()) {
        return None;
    }
    // Must be at top or bottom of page.
    // US Letter = 792pt, A4 = 841pt. Page numbers are typically in the
    // top ~5% or bottom ~12% of the page.
    if item.y <= PAGE_NUMBER_TOP_Y && item.y >= PAGE_NUMBER_BOTTOM_Y {
        return None;
    }
    text.parse().ok()
}
```

Removal happens in `page_number_removal_mask` (`src/extractor/layout.rs:1605-1615`) with no cross-page corroboration:

```rust
.map(|(index, value)| explicit_folio[index] || (value.is_some() && !contextual[index]))
```

The predicate is only "1–4 ASCII digits", so the magnitude is unbounded: on a real document, shifting content upward progressively deletes every index that crosses y=720, including two-digit ones (`10`…`18`).

The `contextual` guard does not appear to engage in this layout. With the synthetic PDF above, a digit flanked by words on the same baseline is still deleted (`Row 1 December` → `Row December`) at every column spacing tested from 8 pt to 24 pt, while the identical row at y=700 is kept.

## The tokens are present — the loss is in markdown assembly

Two independent extractors recover them from the same file:

```
pdftotext (poppler):  n=\n1\n\n22 December 2026\n\n2\n\n22 January 2027 …
pypdf:                'n = ' / '1 22 December 2026 ' / '2 22 January 2027 ' …
```

and `extract_text_with_positions` still reports them with correct coordinates:

```
y=753.78 x=208.4 'n = '
y=738.78 x=208.4 '1 '   x=226.4 '22 December 2026 '
y=723.78 x=208.4 '2 '   x=226.4 '22 January 2027 '
y=708.78 x=208.4 '3 '   x=226.4 '22 February 2027 '   <- survives
```

So the text extraction is fine; the deletion is introduced only on the markdown path.

## Real-world impact

We parse A4 financial documents (EU securities Final Terms and PRIIPs KIDs). Across a 17-document corpus from six issuers, digit-only tokens above y=720 were deleted in six documents. Excluding genuine folios below y=100, the losses were:

| Document | Page | Deleted | What it was |
|---|---|---|---|
| Final Terms A (29 pp) | 20 | `13 14 15 28 29 30` | row indices of an autocall payment schedule |
| " | 17, 24 | `2025 2024 2023` | year column headers in the financial statements |
| Final Terms B (29 pp) | 9 | `4` | a row index in an underlying table |
| KID (3 pp) | 2 | `5` | a step of the 1–7 risk scale |
| Final Terms C (10 pp) | 3, 6 | `1 2` (each) | row indices of two redemption schedules |

The dangerous property is that the emitted markdown is still **structurally valid**:

```
|10 June 2027|87.00%|100.00%|17 June 2027|      <- index 14 gone
|12 July 2027|86.00%|100.00%|19 July 2027|      <- index 15 gone
```

Row count, column count and date ordering all pass. Nothing short of an explicit "indices must be contiguous 1..N" invariant detects it, and that invariant is only writable by someone who already knows about this behaviour.

Whether a given document is affected depends purely on where its table happens to land on the page, so a document set that looks clean today can start losing rows after an unrelated layout change upstream.

## No way to turn it off

`filter_markdown_page_numbers_with_removed_pages` is called unconditionally (`src/lib.rs:489-490`). `MarkdownOptions::include_page_numbers` does not control it — it only decides whether `<!-- Page N -->` comments are emitted (`src/markdown/convert.rs:855`, `:1310`). The Python binding is `process_pdf(path, pages)` and exposes no options at all, so from Python the behaviour cannot be influenced.

## Suggested fixes

1. **Scale the band by the page's actual MediaBox height** (top 5% / bottom 12%, matching the comment) instead of the absolute 720.0 / 100.0 constants. This is the correct fix but appears to need page geometry threaded into `layout.rs`, since `TextItem` carries no page dimensions.

2. **A smaller fix that needs no new data: require a page-number candidate to be alone on its baseline.** A real folio is typically the only item on its line; in every case above the deleted digit shared a baseline with ordinary body text one column to the right. Rejecting candidates that have a neighbour on the same baseline within the existing `PAGE_NUMBER_Y_TOLERANCE` would eliminate these false positives. Requiring corroboration across pages (same x/y, advancing value) before deletion would harden it further.

Even without a fix, exposing the removal as an option on `PdfOptions` and on the Python bindings would let callers opt out.

## Environment

- `pdf-inspector` 0.2.6 (PyPI wheel), Python 3.12.13, macOS 26.5.1 (arm64)
- Source read at `585d36e` ("fix: recover from a corrupted startxref pointer (#230)")

### PR DESCRIPTION
**Pull request by mathurshubham**
Fixes #283 — building directly on @IngLP's excellent diagnosis.

## What

`page_number_value` gated folio candidates on absolute Y constants calibrated
for US Letter (`PAGE_NUMBER_TOP_Y = 720.0` of a 792pt page — the top inch).
On taller pages the band grows past the intended physical margin: on A4
(841.89pt) it covers the top **~122pt, roughly eight lines of body text**, so
short digit-only tokens there — table row indices, year column headers, risk
scale steps — were silently deleted as folios while the output stayed
structurally valid.

This PR scales both bands by the page's visible-box height (CropBox, else
MediaBox, via the existing `get_page_box`), threading a per-page `(y0, y1)`
map from the document into the folio filters:

- **US Letter keeps the exact historical thresholds** (scale = 1.0) — no
  behavior change for the calibrated case, so the existing corpus semantics
  are preserved.
- A4's top band becomes y > 765.35 (the same physical inch), bottom
  y < 106.3; boxes with a nonzero origin shift with it.
- Pages without resolvable geometry — and the item-only grouping path that
  has no document access — keep the absolute constants, so behavior degrades
  to exactly today's, never something new.

## One note on the repro

The issue's synthetic repro (18pt column gap) is already rescued on current
`main` by the baseline-context grouping, but the underlying miscalibration
still fires whenever the digit's neighbor sits beyond the 1.5-em context gap
— which is precisely the reported real-world shape (an index column beside a
percentage column):

```
A4, rows at y=740/725:  "14 | 87.00%"  →  "87.00%"     (index deleted, main)
                                       →  "14 | 87.00%" (this PR)
```

Verified with generated single-page PDFs on both paper sizes:

| Case | main | this PR |
|---|---|---|
| A4, `14` + far column at y=740 (body) | index **deleted** | kept |
| A4, lone `7` at y=810 (real top margin) | stripped | stripped |
| Letter, `14` + far column at y=740 (folio zone) | stripped | stripped (unchanged) |

## Tests

- `page_number_bands_scale_with_page_height` — Letter identity, A4 scaling,
  nonzero-origin boxes, missing/degenerate geometry fallback.
- `a4_body_digits_above_the_letter_band_are_not_candidates` — the issue's
  case; also pins that A4's real top margin still yields candidates and that
  the bottom band scales symmetrically.
- `letter_pages_keep_their_historical_bands` — boundary values (99/101,
  719/740) identical with and without geometry.
- Full suite: 822 lib tests + 152 + integration all pass; `cargo fmt`,
  `clippy -D warnings`, wasm target check/clippy, and the developer-script
  tests are green. No existing test needed its expectations changed —
  Letter-calibrated behavior is untouched by construction.

## Not included (possible follow-ups from the issue)

- The same-baseline-neighbor hardening (suggested fix 2) — the context
  grouping already covers gaps up to 1.5em; widening it is a behavior change
  for Letter documents too, so it deserves its own discussion.
- Exposing the removal as an option on `PdfOptions`/bindings — orthogonal
  API surface, better decided by maintainers.

---
<a href="https://app.blacksmith.sh/firecrawl/codesmith/pdf-inspector/pr/287"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with [code]smith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1788633172&installation_model_id=11126&pr_number=287&repository=firecrawl%2Fpdf-inspector&return_to=https%3A%2F%2Fgithub.com%2Ffirecrawl%2Fpdf-inspector%2Fpull%2F287&signature=cd4121185ed18094e8e0c46e9b0b06861f6e14b78300638b8c53cb3f560591eb"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with [code]smith</picture></a>
<sup>Need help on this PR? Tag <code>@codesmith-bot</code> with what you need. Autofix is disabled.</sup>

---
## Summary by cubic
Scale page-number detection bands by each page’s real height so taller pages (like A4) keep the same physical margins, and apply the same bands to contextual folio checks. This prevents accidental removal of digit-only body text near the top/bottom; US Letter behavior is unchanged and pages without geometry fall back to existing constants.

- **Bug Fixes**
  - Scale top/bottom folio bands using per-page visible-box height; A4 now uses inch-equivalent thresholds.
  - Use the same scaled bands in the deep-margin recurrence gate so repeated folios on tall pages are recognized; fall back to absolute constants when geometry is missing or in item-only paths.
  - Thread per-page `(y0, y1)` into folio filters and selection; add tests for Letter identity, A4 scaling, nonzero origins, missing geometry, and the contextual-gate consistency.

<sup>Written for commit 73d77efa2c3b3887a08fec7491b2a9eef3eef1a0. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/firecrawl/pdf-inspector/pull/287?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

**Comment by princeamir56** [comment]
# PR Review — #287: fix(layout): scale the page-number bands by the page's real height

**Repository**: firecrawl/pdf-inspector
**Author**: @mathurshubham
**Branch**: fix/scale-page-number-band → main
**Date**: 2026-08-07
**Reviewed by**: PR Review Agent (Orchestrator + 3 agents)

---

## 📋 Summary

**What changed:** fix(layout): scale the page-number bands by the page's real height. Touches 4 files (src/) with +192/-26 lines.

**Why it exists:** `bug fix` — title/commits reference a fix or bug.

**Impact areas:** `src/extractor`, `src/lib.rs`, `src/markdown`

**Complexity score:** Medium — 4 files, 218 changed lines, 3 impact areas.

**Files changed summary:**
- `src/extractor/layout.rs` — modified (rs) (+146/-17)
- `src/lib.rs` — modified (rs) (+28/-8)
- `src/extractor/mod.rs` — modified (rs) (+17/-1)
- `src/markdown/mod.rs` — modified (rs) (+1/-0)

---

## 🔒 Security Analysis

**Risk level:** 🟢 Clean

✅ No security issues detected

**Categories checked:** secrets ✅ · injection ✅ · auth ✅ · dependencies ✅ · hardcoded ✅ · input-validation ✅ · sql ✅ · xss ✅ · cors/csp ✅ · path-traversal ✅ · ssrf ✅ · deserialization ✅ · crypto ✅ · open-redirect ✅ · cookie ✅ · tls ✅ · prototype-pollution ✅ · header-injection ✅

**External scanners:** gitleaks, Semgrep OSS, Trivy ran — no findings beyond the rule engine.

_Suppress a false positive by adding `// pr-agent-ignore` or `// nosec` on the flagged line._

---

## 📚 Documentation

**Missing docstrings/JSDoc:**
- None — all detected public symbols are documented.

**README impact:** No README update required.

**Changelog entry:**
```markdown
## [Unreleased]
### Fixed
- fix(layout): scale the page-number bands by the page's real height.
```

**Auto-generated docstrings:**
_No function declarations detected in the diff._

---

## ✅ Review Decision

| Agent | Status | Risk |
|---|---|---|
| Summary | ✅ Complete | — |
| Security | 🟢 Clean | Clean |
| Documentation | 🟢 OK | Low |

**Overall recommendation**: APPROVE

---
*Generated by PR-Review-Agent on 2026-08-07 01:33:21 UTC*

File: src/extractor/layout.rs [hunk 1]
```diff
@@ -2001,7 +2036,7 @@ fn group_into_lines_with_thresholds_and_regions_impl(
         // from changing the page's inferred layout.
         let column_detection_items: Vec<TextItem> = page_items
             .iter()
-            .filter(|item| page_number_value(item).is_none())
+            .filter(|item| page_number_value(item, &HashMap::new()).is_none())
```
[the remark this answers is not in the document]
Comment (mathurshubham) [reply src/extractor/layout.rs:2043]: Deliberately left as-is, and I'd argue it's out of this PR's scope: excluding folio-shaped digits from column inference (while retaining them in output) is a pre-existing pattern on `main` — contextual candidates on Letter pages have always been kept in the output yet hidden from the column histogram by this same filter. With the empty map this call keeps the exact absolute bands it used before this PR, so column detection and reading order are byte-for-byte unchanged for every input. Threading the bounds map here means changing the public `group_into_lines*` family's signatures (it's `pub`, used by `tables/` and the bindings-facing paths), which is a wider API decision than this fix should smuggle in. Happy to do it as a follow-up if maintainers want the grouping APIs geometry-aware.

File: src/extractor/layout.rs [hunk 2]
```diff
@@ -984,10 +1006,11 @@ fn page_number_value(item: &TextItem) -> Option<u32> {
...
     }
 
-    // Must be at top or bottom of page.
-    // US Letter = 792pt, A4 = 841pt. Page numbers are typically in the
-    // top ~5% or bottom ~12% of the page.
-    if item.y <= PAGE_NUMBER_TOP_Y && item.y >= PAGE_NUMBER_BOTTOM_Y {
+    // Must be at the top or bottom of the page. The bands scale with the
+    // page's real height so taller pages (A4 and up) keep the same physical
+    // margins as US Letter instead of swallowing body text.
+    let (bottom, top) = page_number_bands(page_bounds.get(&item.page));
```
[the remark this answers is not in the document]
Comment (mathurshubham) [reply src/extractor/layout.rs:1012]: Valid — fixed in 73d77ef. `page_number_context_masks` now receives the same per-page bounds and the deep-margin gate derives from `page_number_bands`, so contextual candidates inside a taller page's scaled band reach the repeated-folio decision. Added `contextual_deep_margin_gate_uses_the_scaled_bands` pinning the recurring `42 Company report` at A4 y=103 (strips with geometry, stays body without).

**Comment by rmacy** [comment]
Verified at `73d77ef` with an independent, dependency-free A4 fixture.

The fixture places digit-only row indices at y=740 and y=725. Their value column starts far enough away to bypass the existing same-baseline context guard.

Current `main` at `436af97`:

```text
22 December 2026 22 January 2027
```

This PR:

```text
1 22 December 2026 2 22 January 2027
```

Both indices survive, and the order stays correct. This directly covers the reported A4 failure rather than the nearer-column shape that current `main` already protects.

The fixture is synthetic and contains no external document content.

**Comment by IngLP** [comment]
Thank you for picking this up, and for the care in the write-up — the diagnosis and the shape of the fix are right, and keeping US Letter numerically exact is the correct conservative call.

I built the branch and measured it against released 1.14.1 on real documents, and found two cases where scaling costs content. Both are the same failure mode as #283 — structurally valid Markdown, a digit silently gone — just at different page geometries.

**The bottom band grows on tall pages.** On A4 it goes from 100 to 106.3. On the arXiv paper `arXiv:2606.03264v1` that deletes the exponent of a displayed equation: a lone `2` at y=105.52 on page 13, which `extract_text_with_positions` still reports. It is inside the scaled band and outside the calibrated one.

```
main f4aab3b → this branch, process_pdf Markdown, page 13
- *𝐾* 2 1 ∑︁(*𝑘*) *𝑉𝑟*(*𝑥*)= *𝑟* (*𝑥*)−*𝑟*mean(*𝑥*), (3) *𝐾* *𝑘*=1
+ *𝐾* 1 ∑︁(*𝑘*) *𝑉𝑟*(*𝑥*)= *𝑟* (*𝑥*)−*𝑟*mean(*𝑥*), (3) *𝐾* *𝑘*=1
```

The test at `layout.rs` (`page_number_value(bottom@103) == Some(9)` on A4) pins this as intended, so the suite cannot catch it.

**The top band appears on short pages.** This one is quieter. On 1.14.1 the gate is `y > 720`, so on any page shorter than 720pt there is effectively no top band — nothing above y=100 was ever a folio candidate. Scaled, an A5 page (595.28) gets `y > 541.2` and a 16:9 slide gets `y > 490.9`, so digit-only content in the first lines of those pages becomes removable where it was unconditionally safe. No test covers a non-Letter, non-A4 trim.

I have opened #368 with your two commits underneath, unchanged apart from a rebase onto `main`. It keeps your diagnosis and changes what the top band is measured against: `(y1 - PAGE_NUMBER_TOP_MARGIN).max(PAGE_NUMBER_TOP_Y)`, one inch below the page's real top edge, clamped so it can never drop below the calibrated 720. A4 lands at 769.89 rather than 765.35 — same fix, 4.5pt apart — and A5 keeps the unreachable band it always had. The bottom band I ended up not touching at all: every version of it I tried only deleted something. Anchoring it to the box origin has the same failure as scaling it, on a print A4 whose CropBox is inset for trim.

It also carries the geometry to two other places that decide the same question and were left on the absolute constants — column detection, and `try_build_table_from_columns`, both reachable from the document Markdown path — plus an integration test that fails on `main` (loses the top digit), fails on this branch (loses the bottom one), and passes with both clamps.

Happy for it to be merged, cherry-picked, or folded into this PR instead — whatever the maintainers prefer. The measurement I could not do is the one `AGENTS.md` asks for: `pdf-evals` is not public, so the snapshot suite is still unrun for both branches.
